### ISSUE TITLE
df resetting after adding a row

### ISSUE THREAD

**Issue by lalo-ap** [PablocFonseca/streamlit-aggrid#67]
Good day everyone,

I'm trying to create a new df using AgGrid but every time I add a new row, the df resets without saving the inputs previously typed.

Do you know how to avoid this issue?

**Comment by PablocFonseca** [comment]
It seems more an issue of state management. I think you'll need to hold you dataframe in st.session_state. Could you please share some code?

**Comment by NichtJens** [comment]
I had a similar problem, and it is not a problem of session state management as far as I could tell.

I believe it's a potential bug in the behavior depending on whether `reload_data` is enabled/disabled.

To solve it (or maybe it could be considered a workaround, if the behavior is not as intended), I now conditionally set `reload_data`.

Basically, I hold the dataframe in a cache (this could be in the streamlit session state) as well as a boolean flag for whether the dataframe has changed outside of editing cells in AgGrid. This flag I pass in as `reload_data`.

If I don't do that, I am having the same @lalo-ap describes: The dataframe taken from the AgGrid repsonse does not contain the changes made outside of AgGrid. If I now write that dataframe back into the cache, I lose that change.

**Comment by NichtJens** [comment]
Here's a (I hope) minimal example that shows the problem:

```python
import streamlit as st
from st_aggrid import AgGrid
import pandas as pd
​
​
st.write("press \"add row\" button a few times, then edit a cell.")
st.write("try with reload data enabled and disabled.")
​
def mkdf(val):
    return pd.DataFrame.from_records([{"val": val}])
​
df = st.session_state.get("df", mkdf(0))
​
def add_row():
    i = df.get("val").max()
    new = mkdf(i+1)
    st.session_state.df = pd.concat([df, new], ignore_index=True)
​
st.button("add row", on_click=add_row)
​
reload_data = st.checkbox("reload data", value=True, key="reload_data")
​
response = AgGrid(
    df,
    editable=True,
    reload_data=reload_data,
    key="grid"
)
​
st.session_state.df = new_df = response.get("data")
​
c1, c2 = st.columns(2)
​
c1.write("before:")
c1.dataframe(df)
​
c2.write("after:")
c2.dataframe(new_df)
```

**Comment by aduverger** [comment]
@NichtJens I think the best way is to use two different dataframes in cache: the one you pass as input to AG Grid ( for e.g. called `df`), and the one that comes out of `AgGrid.get("data")` (for e.g. called `changed_df`). 
The trick here is that, when you add a row, you concat the new row with `changed_df`, and you assign this concatenation to `df`. In this way, what was previously typed is saved.

Dealing with `True` / `False` for `reload_data` does not seem to be stable.

You could adapt your code with the following:

```python
import pandas as pd
import streamlit as st
from st_aggrid import AgGrid


def mkdf(val):
    return pd.DataFrame.from_records([{"val": val}])


# Initialize session_state
if "df" not in st.session_state:
    st.session_state["df"] = mkdf(0)


def add_row():
    i = st.session_state.changed_df.get("val").max()
    new_row = mkdf(i + 1)
    st.session_state.df = pd.concat(
        [st.session_state.changed_df, new_row], ignore_index=True
    )


st.button("add row", on_click=add_row)

grid = AgGrid(st.session_state.df, editable=True)

st.session_state.changed_df = grid.get("data")
# Cast correct dtype (st.aggrid sometime mess with casting):
st.session_state.changed_df = st.session_state.changed_df.astype(
    {col: st.session_state.df[col].dtypes for col in st.session_state.df.columns}
)

c1, c2 = st.columns(2)

c1.write("before:")
print(st.session_state.df.dtypes)
c1.dataframe(st.session_state.df)

c2.write("after:")
print(st.session_state.changed_df.dtypes)
c2.dataframe(st.session_state.changed_df)

```

**Comment by standard-outlier** [comment]
I am encountering a similar issue where I have two data grids, A and B.  B is a calculated result from A. 

When I select a row in grid A, grid B should update, that works for me.
The problem occurs when I try to allow the user to edit grid B, so I cached the updates to the session. 

Although the cached data is updated, the grid does not reflect the changes. On the other hand, if I set `reload_data=False`, the updated values in grid B are visible, but the grid doesn't refresh when I select another row in grid A.
